File: Ex13.5_Largest_black_subrect/codes/stack.cpp
#include <bits/stdc++.h>
using namespace std;

int largestRectangleHistogram(const vector<int>& h) {
    stack<int> st;
    int maxArea = 0;
    int m = h.size();
    vector<int> H = h;
    H.push_back(0);
    for (int i = 0; i <= m; ++i) {
        while (!st.empty() && H[st.top()] > H[i]) {
            int height = H[st.top()]; st.pop();
            int width = st.empty() ? i : (i - st.top() - 1);
            maxArea = max(maxArea, height * width);
        }
        st.push(i);
    }
    return maxArea;
}

int maximalRectangle(const vector<vector<int>>& A, int n, int m) {
    vector<int> heights(m, 0);
    int best = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            heights[j] = (A[i][j] == 1 ? heights[j] + 1 : 0);
        }
        best = max(best, largestRectangleHistogram(heights));
    }
    return best;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    vector<vector<int>> A(n, vector<int>(m));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < m; ++j)
            cin >> A[i][j];

    cout << maximalRectangle(A, n, m);
    return 0;
}

File: hustack/Ex13.5_Largest_black_subrect/codes/stack.py
def largest_rectangle_histogram(h):
    stack = []
    max_area = 0
    h.append(0)
    for i, height in enumerate(h):
        while stack and h[stack[-1]] > height:
            H = h[stack.pop()]
            W = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, H * W)
        stack.append(i)
    h.pop()
    return max_area

def maximal_rectangle(matrix, n, m):
    heights = [0] * m
    best = 0
    for i in range(n):
        for j in range(m):
            if matrix[i][j] == 1:
                heights[j] += 1
            else:
                heights[j] = 0
        best = max(best, largest_rectangle_histogram(heights))
    return best

if __name__ == "__main__":
    n, m = map(int, input().split())

    A = [[0 for _ in range(m)] for _ in range(n)]

    for i in range(n):
        row = list(map(int, input().split()))
        for j in range(m):
            A[i][j] = row[j]
    
    print(maximal_rectangle(A, n, m))
